events.go

```diff
@@ -0,0 +1,182 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"regexp"
+	"strconv"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	"github.com/google/uuid"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/rs/zerolog"
+	tmrpcclient "github.com/tendermint/tendermint/rpc/client"
+	tmrpceventstream "github.com/tendermint/tendermint/rpc/client/eventstream"
+	rpchttp "github.com/tendermint/tendermint/rpc/client/http"
+	"github.com/tendermint/tendermint/rpc/coretypes"
+	tmjsonclient "github.com/tendermint/tendermint/rpc/jsonrpc/client"
+	tmtypes "github.com/tendermint/tendermint/types"
+)
+
+var (
+	queryEventBankTransfer = fmt.Sprintf(
+		"%s='%s' AND %s='%s'",
+		tmtypes.EventTypeKey,
+		tmtypes.EventTxValue,
+		"message.action",
+		"/cosmos.bank.v1beta1.MsgSend",
+	)
+)
+
+type EventCollector struct {
+	rpcClient tmrpcclient.Client
+	logger    zerolog.Logger
+	counter   *prometheus.CounterVec
+	up        *prometheus.GaugeVec
+}
+
+func NewEventCollector(tmRPC string, logger zerolog.Logger) (*EventCollector, error) {
+	httpClient, err := tmjsonclient.DefaultHTTPClient(tmRPC)
+	if err != nil {
+		return nil, err
+	}
+	// no timeout because we will continue fetching events continuously
+	rpcClient, err := rpchttp.NewWithClient(tmRPC, httpClient)
+	if err != nil {
+		return nil, err
+	}
+	transfersValueCounter := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name:        "cosmos_bank_transfer_amount",
+			Help:        "Number of tokens transferred in a transfer message",
+			ConstLabels: ConstLabels,
+		},
+		[]string{"denom", "sender", "recipient"},
+	)
+	upGauge := prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name:        "cosmos_bank_transfer_up",
+			Help:        "Whether the bank transfer event collector is up",
+			ConstLabels: ConstLabels,
+		},
+		[]string{"denom", "sender", "recipient"},
+	)
+	return &EventCollector{
+		rpcClient: rpcClient,
+		logger:    logger,
+		counter:   transfersValueCounter,
+		up:        upGauge,
+	}, nil
+}
+
+func (s EventCollector) Start(
+	ctx context.Context,
+) error {
+	if err := s.rpcClient.Start(ctx); err != nil {
+		return err
+	}
+	s.logger.Info().Msg("Starting StreamCollector")
+	go s.RunBankTransferEventStream(ctx)
+
+	return nil
+}
+
+func (s EventCollector) RunBankTransferEventStream(ctx context.Context) {
+	eventStream := tmrpceventstream.New(s.rpcClient, queryEventBankTransfer, &tmrpceventstream.StreamOptions{
+		WaitTime: 300 * time.Millisecond,
+	})
+	streamEventErr := make(chan error, 1)
+	go func() {
+		streamEventErr <- eventStream.Run(ctx, s.HandleBankTransferEvent)
+	}()
+	for {
```

Comment:
Should we add a backoff here on the reties?

Reply:
I dont think so, because it only retries when it errors after missing some events, which is known to be recoverable via retry. I believe the main reason eventstream shuts down in that case is so that there isnt an at least once expectation for event consumption.